realm/realm-library/build.gradle

```diff
@@ -259,3 +295,127 @@ artifacts {
+
+        println "Existing archive hash mismatch (Expected: ${project.coreSha256Hash.toLowerCase()}" +
+                " but got ${calculatedHash.toLowerCase()}). Download new version."
+        return true
+    }
+
+    doLast {
+        if (shouldDownloadCore()) {
+            download {
+                src "http://static.realm.io/downloads/core/realm-core-android-${project.coreVersion}.tar.gz"
+                dest project.coreArchiveFile
+                onlyIfNewer false
+            }
+            coreDownloaded = true
+
+            if (isHashCheckingEnabled()) {
+                def calculatedHash = calcSha256Hash(project.coreArchiveFile)
+                if (!project.coreSha256Hash.equalsIgnoreCase(calculatedHash)) {
+                    throw new GradleException("Invalid checksum for file '" +
+                            "${project.coreArchiveFile.getName()}'. Expected " +
+                            "${project.coreSha256Hash.toLowerCase()} but got " +
+                            "${calculatedHash.toLowerCase()}.");
+                }
+            } else {
+                println 'Skipping hash check (empty \'coreSha256Hash\').'
+            }
+        }
+    }
+}
+
+task compileCore(group: 'build setup', description: 'Compile the core library from source code') {
+    // Build the library from core source code
+    doFirst {
+        if (!coreSourcePath) {
+            throw new GradleException('The coreSourcePath is not set.')
+        }
+        exec {
+            workingDir = coreSourcePath
+            commandLine = [
+                    "bash",
+                    "build.sh",
+                    "build-android"
+            ]
+        }
+    }
+
+    // Copy the core tar ball
+    doLast {
+        copy {
+            from "${coreSourcePath}/realm-core-android-${coreVersion}.tar.gz"
+            into project.coreArchiveFile.parent
+            rename "realm-core-android-${coreVersion}.tar.gz", "core-android-${coreVersion}.tar.gz"
+        }
+    }
+}
+
+task deployCore(group: 'build setup', description: 'Deploy the latest version of Realm Core') {
+    dependsOn {
+        coreSourcePath ? compileCore : downloadCore
+    }
+
+    outputs.upToDateWhen {
+        // Clean up the coreDir if it is newly downloaded or compiled from source
+        if (coreDownloaded || coreSourcePath) {
+            return false
+        }
+
+        return project.coreDir.exists()
+    }
+
+    doLast {
+        // Delete all files to avoid multiple copies of the same header file in Android Studio.
+        exec {
+            commandLine = [
+                    'rm',
+                    '-rf',
+                    project.coreDistributionDir.getAbsolutePath()
+            ]
+        }
+        copy {
+            from tarTree(project.coreArchiveFile)
+            into project.coreDir
+        }
+    }
+}
+
+preBuild.dependsOn deployCore
+
+task cleanExternalBuildFiles(type: Delete) {
+    delete project.file('.externalNativeBuild')
```

Comment:
@beeender Can I update `cleanExternalBuildFiles` task to have the task respect `dontCleanJniFiles` property?

The task body should be something like:
```
    if (!project.hasProperty('dontCleanJniFiles')) {
         delete project.file('.externalNativeBuild')
    } else {
        doLast {
            tasks.all { task ->
                if (task.name.startsWith('externalNativeBuildClean')) {
                    task.enabled = false
                }
            }
        }
    }
```